doc/SNS_Overview.rst

```diff
@@ -0,0 +1,73 @@
+
+****************
+Items
+****************
+
+- Striping is, generally, two-level. SNS stores a file in a collection of component files, using logical offsets within component files to determine where file data go. A component file is stored by a C2 back-end in a container, which is a thin abstraction layer on top of device. A container can itself be striped over other containers (or de-duped, encrypted, compressed, etc.). 
+
+- The same algorithms and the same data-structures are used for SNS and local RAID.
+
+- The key piece of meta-data used to implement striping (and all other forms of file storage in C2) is layout. A layout is a map that transforms an offset in a file into a location where data at this offset reside. For example, an SNS file layout transforms a file offset into a (component-file-id, offset) pair, where offset is a logical offset within the component file. A component file has layout too, and this layout can be used to transform the offset into location. In other words, layouts are traversed until final data location is found. 
+
+- As we shall see, a layout can be rather complex and storing large layout as an attribute of every file can be prohibitively expensive (compare with Lustre LOVEA story). To address this, a notion of layout formula is introduced. A layout formula takes some parameters and produces a layout as result. Instead of storing a file layout together with the file, the file references a layout formula shared by a large number of files. To obtain file layout, a few parameters, including fid, are substituted into the formula. For example, instead of layout L = {RAID0 over component files  with fids CF0, CF1, ..., CF127 } a formula LFun(n, f, s) = {RAID0 over component files with fids f, f + 1*s, f + 2*s, ..., f + (n-1)*s} can be used. A particular file layout can then be described as LFun(128, fid0, step0). 
+
+- Design of layout formulae is a critical condition of efficient striping implementation. The class of formulae must conform to the following requirements: 
+
+  - all standard RAID patterns must be supported compactly; 
+
+  - parity declustering must (see below) be supported; 
+
+  - formulae should work efficiently with small FLDB (Fid Location Data-Base) and CLDB (Container Location Data-Base), in the face of object migration. 
+  
+- Concurrency control. Concurrent conflicting operations on a striped file must preserve consistency. Striping with redundancy introduces dependencies between data updates: parity and corresponding data must be updated atomically. T1 does not provide concurrent access to exported block devices, but still have some concurrency, e.g., during recovery.
```

Comment:
Replace "T1" with "Current version".

Do the same for other places.

Reply:
Incorporated.